python/mxnet/optimizer.py

```diff
@@ -349,26 +351,12 @@ def update(self, index, weight, grad, state):
         if self.clip_gradient is not None:
             grad = clip(grad, -self.clip_gradient, self.clip_gradient)
 
-        if state:
-            mom = state
-            mom[:] *= self.momentum
-            if self.weight_previous.has_key(index):
-                mom[:] += -lr * (grad + wd * weight + self.lamda \
-                                    * grad * grad * (weight - self.weight_previous[index]))
-                self.weight_previous[index] = weight
-            else:
-                mom[:] += -lr * (grad + wd * weight)
-                self.weight_previous[index] = weight
-            weight[:] += mom
-        else:
-            assert self.momentum == 0.0
-            if self.weight_previous.has_key(index):
-                weight[:] += -lr * (grad + wd * weight + self.lamda \
-                                    * grad * grad * (weight - self.weight_previous[index]))
-                self.weight_previous[index] = weight
-            else:
-                weight[:] += -lr * (grad + wd * weight)
-                self.weight_previous[index] = weight
+        mom, previous_weight = state
```

Comment:
This wastes memory and compute. better use different states and test for momentum == 0